{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8b7635e1-1fca-4390-a3e9-f56c10bb20ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/liuyi/miniconda3/envs/spatialsnv/lib/python3.10/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['font.sans-serif'] = 'Arial'\n",
    "# pip install pysam\n",
    "import pysam\n",
    "# pip install concave_hull\n",
    "import concave_hull\n",
    "\n",
    "root_path = \"/storage/liuyi/00.SNP_project/code/gigascience_update\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253db9bd-c78b-4511-8876-7d38c1dd5ca3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# bulid mutpep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbeb10c7-39d1-4f51-b0b6-133e1198f141",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene2symble = pd.read_csv(f\"{root_path}/otherdata_for_figure/Figure6/gene2ensembl.gz\",sep = '\\t',compression='gzip')\n",
    "gene2pep = dict(zip(gene2symble['Ensembl_gene_identifier'],gene2symble['protein_accession.version']))\n",
    "tran2pep = dict(zip(gene2symble['Ensembl_rna_identifier'],gene2symble['protein_accession.version']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62d76c6e-27e8-45b6-800a-17ea95fd73db",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "def cds_fa_protein(file_path):\n",
    "    fasta_dict = {}\n",
    "    current_protein_id = None\n",
    "    current_sequence = []\n",
    "\n",
    "    with gzip.open(file_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if current_protein_id:\n",
    "                    fasta_dict[current_protein_id] = ''.join(current_sequence)\n",
    "                header_parts = line.split()\n",
    "                for part in header_parts:\n",
    "                    if part.startswith('[protein_id='):\n",
    "                        current_protein_id = part.split('=')[1][:-1]\n",
    "                        break\n",
    "                current_sequence = []\n",
    "            else:\n",
    "                current_sequence.append(line)\n",
    "        if current_protein_id:\n",
    "            fasta_dict[current_protein_id] = ''.join(current_sequence)\n",
    "    \n",
    "    return fasta_dict\n",
    "\n",
    "import re\n",
    "def extract_letter_and_number(mutation):\n",
    "    match = re.match(r'([A-Za-z])(\\d+)([A-Za-z])', mutation)\n",
    "    if match:\n",
    "        first_letter = match.group(1)\n",
    "        number = int(match.group(2))\n",
    "        last_letter = match.group(3)\n",
    "        return first_letter, number, last_letter\n",
    "    \n",
    "import os\n",
    "import itertools\n",
    "\n",
    "def SelectPEP(pep_list, prefix):\n",
    "    select_IEDB = linear_epitope.loc[linear_epitope['Epitope_Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_IEDB.shape[0] > 0: select_IEDB.to_csv(\"{}.search_result.IEDB.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "    select_TCell_epitode = Linear_peptide_T_cell.loc[Linear_peptide_T_cell['Epitope_Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_TCell_epitode .shape[0] > 0: select_TCell_epitode.to_csv(\"{}.search_result.IEDB_T_Cell.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "    select_HLALA = Ligand_HLALA.loc[Ligand_HLALA['Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_HLALA.shape[0] > 0: select_HLALA.to_csv(\"{}.search_result.HLA_Ligand_Atlas.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "    select_HLAtlas_Cryptic = Ligand_HLALA_cryptic.loc[Ligand_HLALA_cryptic['Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_HLAtlas_Cryptic.shape[0] >0:select_HLAtlas_Cryptic.to_csv(\"{}.search_result.HLA_Ligand_Atlas_Cryptic.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "    select_IEAtlas = Ligand_IEAtlas.loc[Ligand_IEAtlas['Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_IEAtlas.shape[0] >0: select_IEAtlas.to_csv(\"{}.search_result.IEAtlas.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "    select_dbPepNeo = Ligand_dbPepNeo.loc[Ligand_dbPepNeo['Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_dbPepNeo.shape[0] >0: select_dbPepNeo.to_csv(\"{}.search_result.dbPepNeo.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "    select_MHCMotifAtlas = Ligand_MHC_Motif_Atlas.loc[Ligand_MHC_Motif_Atlas['Sequence_I2L'].isin(pep_list)].copy()\n",
    "    if select_MHCMotifAtlas.shape[0] >0: select_MHCMotifAtlas.to_csv(\"{}.MHC.Motif.Atlas.tsv\".format(prefix), sep=\"\\t\", header=True, index=False)\n",
    "\n",
    "    return([select_IEDB, select_TCell_epitode, select_HLALA, select_HLAtlas_Cryptic, select_IEAtlas, select_dbPepNeo, select_MHCMotifAtlas])\n",
    "\n",
    "def search_database(seq_list,sample,outDir):\n",
    "\n",
    "    DB_path = f\"{root_path}/otherdata_for_figure/Figure6/neoantigen_database\"\n",
    "\n",
    "    global Ligand_IEAtlas\n",
    "    Ligand_IEAtlas = pd.read_csv(os.path.join(DB_path, \"IEAtlas_epitope.I2L.Seq.tsv\"), header=0, sep='\\t',low_memory=False)\n",
    "    global Linear_peptide_T_cell\n",
    "    Linear_peptide_T_cell = pd.read_csv(os.path.join(DB_path, \"IEDB_TCell_Linear_epitope.tsv\"), header=0, sep='\\t',low_memory=False)\n",
    "    global linear_epitope\n",
    "    linear_epitope = pd.read_csv(os.path.join(DB_path, \"IEDB_Linear_epitope.tsv\"), header=0, sep='\\t',low_memory=False)\n",
    "    global Ligand_HLALA\n",
    "    Ligand_HLALA = pd.read_csv(os.path.join(DB_path, \"HLA_Ligand.I2L.Seq.tsv\"), header=0, sep='\\t',low_memory=False)\n",
    "    global Ligand_HLALA_cryptic\n",
    "    Ligand_HLALA_cryptic = pd.read_csv(os.path.join(DB_path, \"HLA_Ligand.cryptic.I2L.Seq.tsv\"), header=0, sep='\\t',low_memory=False)\n",
    "    global Ligand_dbPepNeo\n",
    "    Ligand_dbPepNeo = pd.read_csv(os.path.join(DB_path, \"dbPepNeo.I2L.seq.tsv\"), header=0, sep='\\t', low_memory=False)\n",
    "    global Ligand_MHC_Motif_Atlas\n",
    "    Ligand_MHC_Motif_Atlas = pd.read_csv(os.path.join(DB_path, \"MHC.Motif.Atlas_data_classI_all_peptides.txt\"), header=0, sep=\"\\t\", low_memory=False)\n",
    "    Ligand_MHC_Motif_Atlas['Sequence_I2L'] = Ligand_MHC_Motif_Atlas['Peptide'].str.replace('I', 'L')\n",
    "    \n",
    "    Database = [linear_epitope['Epitope_Sequence_I2L'],\n",
    "                Linear_peptide_T_cell['Epitope_Sequence_I2L'],\n",
    "                Ligand_HLALA['Sequence_I2L'],\n",
    "                Ligand_HLALA_cryptic['Sequence_I2L'],\n",
    "                Ligand_IEAtlas['Sequence_I2L'],\n",
    "                Ligand_dbPepNeo[\"Sequence_I2L\"],\n",
    "                Ligand_MHC_Motif_Atlas['Sequence_I2L']]\n",
    "\n",
    "    DB_length = len(set(itertools.chain(*Database)))\n",
    "    log_text = list()\n",
    "    log_text.append(\"Parsing Public Database Sequnece number: {:d}\".format(DB_length))\n",
    "\n",
    "    # parse input file\n",
    "    # seq_df = pd.read_csv(seq_list,sep='\\t',header=None)\n",
    "    seq_df = pd.DataFrame(seq_list)\n",
    "    seq_df.columns = ['Sequence']\n",
    "    seq_df['Sequnece_I2L'] = seq_df.Sequence.str.replace('I','L')\n",
    "    seq_df = seq_df.drop_duplicates(subset=['Sequence'])\n",
    "\n",
    "    # check if There are matched record\n",
    "    seq_in_num = len(set(seq_df.Sequence.to_list()))\n",
    "    seq_matched_num = len(sorted(set(seq_df.Sequnece_I2L).intersection(set(itertools.chain(*Database)))))\n",
    "    if seq_matched_num == 0:\n",
    "        log_text.append(\"No Sequence were Found in All DataBases.\")\n",
    "    else:\n",
    "        log_text.append(\"Matched number of input Sequence ({:d}):\\n\\t{:d}({:.2f}%)\".format(seq_in_num, seq_matched_num, seq_matched_num/seq_in_num*100))\n",
    "\n",
    "        Included_DB = SelectPEP(seq_df.Sequnece_I2L.to_list(), os.path.join(outDir, sample))\n",
    "        search_status = seq_df.copy()\n",
    "        DB_name = ['IEDB', 'IEDB_T_Cell', 'HLA_Ligand_Atlas', 'HLAtlas_Cryptic', 'IEAtlas', 'dbPepNeo','MHC_Motif_Atlas']\n",
    "        for i,n in enumerate(Included_DB):\n",
    "            log_text.append(\"\\t{:d}\\trecords found from\\t{}\\t\".format(n.shape[0], DB_name[i]))\n",
    "            if 'Sequence_I2L' in n.columns:\n",
    "                juge_list = n.Sequence_I2L.to_list()\n",
    "            else:\n",
    "                juge_list = n.Epitope_Sequence_I2L.to_list()\n",
    "            search_status.loc[:,DB_name[i]] = search_status.Sequnece_I2L.apply(lambda x: 1 if x in juge_list else 0)\n",
    "\n",
    "        search_status.to_csv(os.path.join(outDir, \"{}.DB_search.stat\".format(sample)), header=True, index=False, sep=\"\\t\")\n",
    "\n",
    "    f = open(os.path.join(outDir, \"{}.DB_search.log\".format(sample)), 'w')\n",
    "    f.write(\"\\n\".join(log_text))\n",
    "    f.write(\"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d7cea11-2ca2-4e08-a551-cd7c8d3b9058",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdsaadict = cds_fa_protein(f'{root_path}/otherdata_for_figure/Figure6/GCF_000001405.38_GRCh38.p12_translated_cds.faa.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff913642-1817-4924-a678-d382d52c4a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "window_pepdf_dict = {}\n",
    "for sample in ['CRC-P59-T_2','CRC-P59-T_1','CRC-P19-T','CRC-P67-T']:\n",
    "\n",
    "    rna = sc.read_h5ad(f'{root_path}/rna_adata/{sample}.rna.h5ad')\n",
    "    snv = sc.read_h5ad(f'{root_path}/snv_adata/{sample}.snv.h5ad')\n",
    "    snv.X = snv.layers['norm']\n",
    "    annovar = pd.read_csv(f'{root_path}/otherdata_for_figure/Figure6/{sample}.homo_multianno.txt',sep = '\\t')\n",
    "    annovar['snv'] = annovar['Chr']+'_'+annovar['Start'].map(str)+\":\"+annovar['Ref'] + \">\"+annovar['Alt']\n",
    "    aadict = dict(zip(annovar['snv'],annovar['AAChange.gencodev38']))\n",
    "    snv.var['AA'] = snv.var['gene_ids'].map(aadict)\n",
    "    nonsyaa = snv[:,snv.var['ExonicFunc']=='nonsynonymous SNV']\n",
    "    all_list = []\n",
    "    for idx,row in nonsyaa.var.iterrows():\n",
    "        annovarinfos = row['AA']\n",
    "        snv_name = row['gene_ids']\n",
    "        transe_state = False\n",
    "        dna_state = False\n",
    "        gene_result = None\n",
    "\n",
    "        aa_change_list = []\n",
    "        seq_result = []\n",
    "        mut_result = []\n",
    "        from_list = []\n",
    "        mut_pos = []\n",
    "        rest_mut_pos = []\n",
    "\n",
    "        psudo_mut = []\n",
    "\n",
    "        aapos_list = []\n",
    "        for annovarinfo in annovarinfos.split(','):\n",
    "            gene_id,tran_id,exonid,cDNAchange,pchange = annovarinfo.split(':')\n",
    "            aa,aa_pos,aa_after = extract_letter_and_number(pchange.split('.')[1])\n",
    "            aapos_list.append(aa_pos)\n",
    "            gene_id = gene_id.split('.')[0]\n",
    "            trans_result = tran2pep.get(tran_id,False)\n",
    "            if trans_result:\n",
    "                cdsaa = cdsaadict.get(trans_result,False)\n",
    "                if cdsaa and len(cdsaa)> aa_pos:\n",
    "                    if cdsaa[aa_pos-1] == aa:\n",
    "                        transe_state = True\n",
    "                        seq_result.append(cdsaa[max(aa_pos-15,0):aa_pos+14])\n",
    "\n",
    "                        before_seq = cdsaa[max(aa_pos-15,0):aa_pos-1]\n",
    "                        before_len = len(before_seq)\n",
    "                        after_seq = cdsaa[aa_pos:aa_pos+14]\n",
    "                        after_len = len(after_seq)\n",
    "\n",
    "                        mut_result.append(before_seq + aa_after + after_seq)\n",
    "                        mut_pos.append(before_len)\n",
    "                        rest_mut_pos.append(after_len)\n",
    "\n",
    "                        aa_change_list.append(f'{aa}>{aa_after}')\n",
    "                        from_list.append(tran_id)\n",
    "\n",
    "                        if before_len != 14:\n",
    "                            before_seq = '@' * (14 - before_len) + before_seq\n",
    "                        if after_len != 14:\n",
    "                            after_seq = after_seq + '@' * (14 - after_len)\n",
    "\n",
    "                        psudo_mut.append(before_seq + aa_after + after_seq)\n",
    "\n",
    "\n",
    "\n",
    "        if not transe_state:\n",
    "            gene_result = gene2pep.get(gene_id,False)\n",
    "            if gene_result:\n",
    "                cdsaa = cdsaadict.get(gene_result,False)\n",
    "                if cdsaa:\n",
    "                    for aa_pos in aapos_list:\n",
    "                        if len(cdsaa) <= aa_pos:\n",
    "                            continue\n",
    "                        if cdsaa[aa_pos-1]==aa:\n",
    "                            transe_state = True\n",
    "                            seq_result.append(cdsaa[max(aa_pos-15,0):aa_pos+14])\n",
    "\n",
    "                            before_seq = cdsaa[max(aa_pos-15,0):aa_pos-1]\n",
    "                            before_len = len(before_seq)\n",
    "                            after_seq = cdsaa[aa_pos:aa_pos+14]\n",
    "                            after_len = len(after_seq)\n",
    "\n",
    "                            mut_result.append(before_seq + aa_after + after_seq)\n",
    "                            mut_pos.append(before_len)\n",
    "                            rest_mut_pos.append(after_len)\n",
    "\n",
    "                            aa_change_list.append(f'{aa}>{aa_after}')\n",
    "                            from_list.append(tran_id)\n",
    "\n",
    "                            if before_len != 14:\n",
    "                                before_seq = '@' * (14 - before_len) + before_seq\n",
    "                            if after_len != 14:\n",
    "                                after_seq = after_seq + '@' * (14 - after_len)\n",
    "\n",
    "                            psudo_mut.append(before_seq + aa_after + after_seq)\n",
    "\n",
    "        tmpdf = pd.DataFrame({'snv':[snv_name]*len(seq_result),'refaa':seq_result,'mutaa':mut_result,'aachange':aa_change_list,'pos_0base':mut_pos,'rest':rest_mut_pos,'psudo_mut':psudo_mut,'from':from_list})\n",
    "        all_list.append(tmpdf)\n",
    "    mutpepdf = pd.concat(all_list)\n",
    "    mutpepdf = mutpepdf.drop_duplicates(['snv','refaa','mutaa','aachange'],ignore_index=True)\n",
    "    window_pepdf = []\n",
    "    for idx,row in mutpepdf.iterrows():\n",
    "        snv_name = row['snv']\n",
    "        seq = row['psudo_mut']\n",
    "        windows_slide = []\n",
    "        for window_size in range(8,16):\n",
    "            for num in range(15-window_size,15):\n",
    "                candidae_seq = seq[num:num+window_size]\n",
    "                candidae_seq = candidae_seq.replace(\"@\", \"\")\n",
    "                if len(candidae_seq) == window_size:\n",
    "                    windows_slide.append(candidae_seq)\n",
    "        windows_slidedf = pd.DataFrame(windows_slide,columns = ['pep'])\n",
    "        windows_slidedf['snv'] = snv_name\n",
    "        window_pepdf.append(windows_slidedf)\n",
    "    window_pepdf = pd.concat(window_pepdf)\n",
    "#     window_pepdf['pep'].to_csv(f'/storage/liuyi/00.SNP_project/new_article/Figure/Figure5/window_pepdf/{sample}.mutpep',index=None,header=None)\n",
    "#     window_pepdf.to_csv(f'/storage/liuyi/00.SNP_project/new_article/Figure/Figure5/window_pepdf/{sample}.mutpepdf.csv')\n",
    "    search_database(window_pepdf['pep'], sample = sample, outDir = f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search')\n",
    "    window_pepdf_dict[sample] = window_pepdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5169588-fcbe-4ffb-a425-0ce13cfa91c5",
   "metadata": {},
   "source": [
    "# collect mutpep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e8720bc-2992-41f7-8a59-57c86041a686",
   "metadata": {},
   "source": [
    "## collect pep from CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "579fa634-a4d9-4c65-add8-72cd64df1634",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_normal_pep = pd.read_csv(f\"{root_path}/otherdata_for_figure/Figure6/GCF_000001405.38_GRCh38.p12_translated_cds.faa.8_15kmer.csv.gz\",header=None,compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa64f65d-68e0-4e0c-b3e0-3173979e783c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_db(sample):\n",
    "\n",
    "    file_info = {\n",
    "        f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search/{sample}.search_result.IEDB.tsv': ('Epitope_Sequence', 'IEDB', 'Epitope_Parent Organism', 'Homo sapiens'),\n",
    "        f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search/{sample}.search_result.IEDB_T_Cell.tsv': ('Epitope_Sequence', 'IEDB_T', None, None),\n",
    "        f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search/{sample}.search_result.dbPepNeo.tsv': ('Mut_peptide', 'dbPepNeo', None, None),\n",
    "        f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search/{sample}.search_result.HLA_Ligand_Atlas.tsv': ('peptide_sequence', 'HLA_Ligand_Atlas', None, None),\n",
    "        f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search/{sample}.search_result.IEAtlas.tsv': ('Sequence', 'IEAtlas', None, None),\n",
    "        f'{root_path}/otherdata_for_figure/Figure6/neoantigen_search/{sample}.MHC.Motif.Atlas.tsv': ('Peptide', 'MHC_Motif_Atlas', None, None),\n",
    "    }\n",
    "    \n",
    "    dfs = []\n",
    "    for file_path, (col_name, label, filter_col, filter_value) in file_info.items():\n",
    "        if os.path.exists(file_path):\n",
    "            df = pd.read_csv(file_path, sep='\\t')\n",
    "            if filter_col and filter_value:\n",
    "                df = df[df[filter_col] == filter_value]\n",
    "            dfs.append(pd.DataFrame(zip(df[col_name], [label] * df.shape[0])))\n",
    "    \n",
    "    total_search = pd.concat(dfs)\n",
    "    total_search.columns = ['pep','db']\n",
    "    hla_remove = total_search[total_search['db']=='HLA_Ligand_Atlas']\n",
    "    mutpepdf = total_search[total_search['db']!='HLA_Ligand_Atlas']\n",
    "    mutpepdf = mutpepdf[~mutpepdf['pep'].isin(hla_remove['pep'])].copy()\n",
    "    denor_mutpepdf = mutpepdf[~mutpepdf['pep'].isin(total_normal_pep[0])].copy()\n",
    "    return denor_mutpepdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6575a6b3-59fb-4930-a07b-a8c940c04fb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "denor_mutpepdf_dict = {}\n",
    "pep2snvresult = {}\n",
    "snvresult = {}\n",
    "snv_adata = {}\n",
    "for sample in ['CRC-P59-T_2','CRC-P59-T_1','CRC-P67-T']:\n",
    "    denor_mutpepdf = search_db(sample)\n",
    "    denor_mutpepdf = denor_mutpepdf.drop_duplicates(['pep','db'])\n",
    "    denor_mutpepdf['sample'] = sample\n",
    "    denor_mutpepdf_dict[sample] = denor_mutpepdf\n",
    "    window_pepdf = window_pepdf_dict[sample]\n",
    "    tmpresult = window_pepdf[window_pepdf['pep'].isin(denor_mutpepdf['pep'])]\n",
    "    pep2snvresult[sample] = tmpresult\n",
    "    \n",
    "    rna = sc.read_h5ad(f'{root_path}/rna_adata/{sample}.rna.h5ad')\n",
    "    snv = sc.read_h5ad(f'{root_path}/snv_adata/{sample}.snv.h5ad')\n",
    "    snv.obs['cluster'] = rna.obs['cluster']\n",
    "    annovar = pd.read_csv(f'{root_path}/otherdata_for_figure/Figure6/{sample}.homo_multianno.txt',sep = '\\t')\n",
    "    annovar['snv'] = annovar['Chr']+'_'+annovar['Start'].map(str)+\":\"+annovar['Ref'] + \">\"+annovar['Alt']\n",
    "    aadict = dict(zip(annovar['snv'],annovar['AAChange.gencodev38']))\n",
    "    snv.var['AA'] = snv.var['gene_ids'].map(aadict)\n",
    "\n",
    "    snv_adata[sample] = snv\n",
    "    \n",
    "    sub_snv = snv[:,snv.var['gene_ids'].isin(tmpresult['snv'])].copy()\n",
    "    sub_snv.var['AA'] = sub_snv.var['AA'].map(lambda x :  list(set([ i.split(':')[4] for i in x.split(',')])))\n",
    "    snvresult[sample] = sub_snv.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caed2ee1-c939-4185-85cd-701bcb3cba0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "statdf = pd.concat([denor_mutpepdf_dict['CRC-P67-T'],denor_mutpepdf_dict['CRC-P59-T_1'],denor_mutpepdf_dict['CRC-P59-T_2']])\n",
    "statdf['values'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82bb8fa7-da28-4526-b982-978858c9ef59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_heatmap(df,colors = 'Set2',figsize = (20,1.5),save = None):\n",
    "    import matplotlib.colors as mcolors\n",
    "    def generate_color(n, cmap='Set1'):\n",
    "        colormap = plt.get_cmap(cmap)\n",
    "        color_dict = {i: mcolors.rgb2hex(colormap(i / n)) for i in range(1,n+1)}\n",
    "        return color_dict\n",
    "    \n",
    "    if isinstance(colors,list):\n",
    "        colors = dict(zip(range(len(colors)),colors))\n",
    "    elif isinstance(colors,str):\n",
    "        colors = generate_color(df.shape[0],cmap = colors)\n",
    "        colors.update({0:'white'})\n",
    "        \n",
    "    data = df.values\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "    for i in range(data.shape[0]):\n",
    "        for j in range(data.shape[1]):\n",
    "            ax.add_patch(plt.Rectangle(\n",
    "                (j, data.shape[0] - i - 1), 1, 1,\n",
    "                facecolor=colors[data[i, j] * (i + 1)],\n",
    "                edgecolor='black', linewidth=1\n",
    "            ))\n",
    "\n",
    "    ax.set_xlim(0, data.shape[1])\n",
    "    ax.set_ylim(0, data.shape[0])\n",
    "    \n",
    "    ax.set_xticks(np.arange(data.shape[1]) + 0.5)\n",
    "    ax.set_yticks(np.arange(data.shape[0]) + 0.5)\n",
    "    \n",
    "    ax.set_xticklabels(['' for _ in range(data.shape[1])])\n",
    "    \n",
    "    ax.set_yticklabels(df.index[::-1])\n",
    "    \n",
    "    ax.set_aspect('equal')\n",
    "    ax.tick_params(left=False, bottom=False)\n",
    "    if save != None:\n",
    "        plt.savefig(save,bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1586e2ae-14ec-47b0-b111-732b9b87078f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.pivot_table(data = statdf, index = 'pep',columns='sample',values = 'values').fillna(0)\n",
    "data_copy = data.copy()\n",
    "data_copy['sum'] = np.sum(data_copy,axis = 1)\n",
    "data_copy['subsum'] = np.sum(data_copy[['CRC-P59-T_1','CRC-P59-T_2']],axis = 1)\n",
    "data_copy = data_copy.sort_values(['sum','subsum'],ascending=False)\n",
    "data = data.loc[data_copy.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3787a0-af59-4302-9760-d711a97785fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_heatmap(data.T,colors = ['white','#EAC2B4','#F2E0C7','#F5EDBE'],figsize = (30,5))#save = './MainFigure/sample.hm.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9364fb73-9207-487f-9ea9-e0eb06f47454",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbstat = pd.pivot_table(data = statdf, index = 'pep',columns='db',values = 'values').fillna(0)\n",
    "dbstat['HLA_Ligand_Atlas'] = 0\n",
    "dbstat['PepFromCDS'] = 0\n",
    "dbstat = dbstat.loc[data_copy.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cbfa737-0808-4199-aa4d-e602bbddfda0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_heatmap(dbstat.T,colors = ['white']+['#AB2D3E']*7,figsize = (30,1,5))#save = './MainFigure/db.hm.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2ac22a7-ffe9-4df0-b0d7-a5c4b9d834eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "exonic_df = []\n",
    "for sample in ['CRC-P59-T_1','CRC-P59-T_2','CRC-P67-T']:\n",
    "    exonic_tmp_df = snv_adata[sample].var['ExonicFunc'].value_counts().to_frame().reset_index()\n",
    "    exonic_tmp_df['sample'] = sample\n",
    "    exonic_df.append(exonic_tmp_df)\n",
    "exonic_df = pd.concat(exonic_df,ignore_index= True)\n",
    "exonic_df = exonic_df[exonic_df['index'] != '.']\n",
    "exonic_df['index'] = exonic_df['index'].astype(str)\n",
    "exonic_pivot_df = exonic_df.pivot_table(index='index', columns='sample', values='ExonicFunc', aggfunc='sum').fillna(0)\n",
    "exonic_pivot_df = exonic_pivot_df[['CRC-P59-T_2','CRC-P59-T_1','CRC-P67-T']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8f0f85c-37d6-4d45-aa4c-656fea0b23d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = {\n",
    "    'nonsynonymous SNV': '#C6133B',\n",
    "    'synonymous SNV': '#F7A24F',\n",
    "    'startloss': '#90162D',\n",
    "    'stopgain': '#996E2E',\n",
    "    'stoploss': '#EECDC0',\n",
    "    'unknown': '#93A5CB'\n",
    "}\n",
    "\n",
    "ax = exonic_pivot_df.T.plot(kind='bar', stacked=True, figsize=(3, 4), color=[colors[col] for col in exonic_pivot_df.index],width = 0.9)\n",
    "\n",
    "plt.xlabel('Sample')\n",
    "plt.ylabel('Count')\n",
    "plt.title('Stacked Bar Plot of Counts by Sample and Exonic Function')\n",
    "plt.legend(title='Exonic Function', bbox_to_anchor=(1.05, 0.5), loc='center left')\n",
    "plt.tight_layout()\n",
    "# plt.savefig(f'./MainFigure/exonic.bar.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "871227b9-6f4e-44bd-9693-9723c5bb10e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "snv_detail_list = []\n",
    "for sample in ['CRC-P59-T_1','CRC-P59-T_2','CRC-P67-T']:\n",
    "    snv_detail_list.append(set(snv_adata[sample][:,snv_adata[sample].var['ExonicFunc'] == 'nonsynonymous SNV'].var_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b059499b-fbb2-45b3-8806-aef3e141ecb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_venn import venn3\n",
    "venn = venn3(snv_detail_list,['CRC-P59-T_1','CRC-P59-T_2','CRC-P67-T'])\n",
    "venn.get_patch_by_id('100').set_color('#FCF1BA')\n",
    "venn.get_patch_by_id('010').set_color('#F9E2C6')\n",
    "venn.get_patch_by_id('001').set_color('#F6C4B5')\n",
    "venn.get_patch_by_id('110').set_color('#F9DC92')\n",
    "venn.get_patch_by_id('101').set_color('#F3B18F')\n",
    "venn.get_patch_by_id('011').set_color('#F5C787')\n",
    "venn.get_patch_by_id('111').set_color('#F4BA6F')\n",
    "\n",
    "venn.get_patch_by_id('100').set_alpha(1)\n",
    "venn.get_patch_by_id('010').set_alpha(1)\n",
    "venn.get_patch_by_id('001').set_alpha(1)\n",
    "venn.get_patch_by_id('110').set_alpha(1)\n",
    "venn.get_patch_by_id('101').set_alpha(1)\n",
    "venn.get_patch_by_id('011').set_alpha(1)\n",
    "venn.get_patch_by_id('111').set_alpha(1)\n",
    "\n",
    "# plt.savefig('./MainFigure/veen.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "744adafb-4616-40b9-b69b-263c4dfda96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stat =[]\n",
    "for sample in ['CRC-P59-T_1','CRC-P59-T_2','CRC-P67-T']:\n",
    "    gene_name = snvresult[sample]['gene_name'].to_frame()\n",
    "    gene_name['sample'] = sample\n",
    "    gene_stat.append(gene_name)\n",
    "gene_stat = pd.concat(gene_stat)\n",
    "gene_stat['gene_name'] = gene_stat['gene_name'].map(lambda x : x.split(';')[0])\n",
    "gene_stat = gene_stat.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "651c810f-7d5f-45bb-8f20-9c0f3a9482eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = {'CRC-P67-T': '#EEC6B8', 'CRC-P59-T_1': '#FAF1C0', 'CRC-P59-T_2': '#F5E3C9'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56bb2410-b170-4f15-a3f7-4d630fd1ca97",
   "metadata": {},
   "outputs": [],
   "source": [
    "pivot_df = gene_stat.pivot_table(index='gene_name', columns='sample', aggfunc='size', fill_value=0)\n",
    "pivot_df['total'] = pivot_df.sum(axis=1)\n",
    "pivot_df = pivot_df.sort_values(by='total', ascending=False)\n",
    "pivot_df = pivot_df.drop(columns='total')\n",
    "pivot_df.plot(kind='bar', stacked=True, color=[colors[col] for col in pivot_df.columns], figsize=(3.5, 5),width = 0.8,edgecolor = 'black')\n",
    "plt.xlabel('Gene Name')\n",
    "plt.ylabel('Count')\n",
    "plt.legend(title='Sample')\n",
    "plt.xticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "plt.ylim(0,3.5)\n",
    "plt.xticks(rotation=45) \n",
    "# plt.savefig('./MainFigure/gene_bar.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3821d386-c838-43e8-867a-bab216d752c2",
   "metadata": {},
   "source": [
    "# binding Affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13447890-b522-4439-a8a1-70b59f02e96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## from T1K (default) https://github.com/mourisl/T1K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b19cc078-29bf-4cdd-95fe-a8cddd56ad12",
   "metadata": {},
   "outputs": [],
   "source": [
    "badf = pd.read_csv(f'{root_path}/otherdata_for_figure/Figure6/ba.tsv',sep = '\\t')\n",
    "badf = badf.loc[(badf['A1101']+badf['C0702']).sort_values(ascending=False).index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fa8c19e-3758-4456-9732-6d117cfe5ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize = (0.5,5))\n",
    "sns.barplot(data = badf, y= 'Total',x = 'A1101',color = '#00A1DF',ax = ax)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e0f477a-2e53-45f2-a149-92e7a4ad5b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize = (0.5,5))\n",
    "sns.barplot(data = badf, y= 'Total',x = 'C0702',color = '#FAC319',ax = ax)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3698a1a-21f5-482b-b153-bab29bf8d88e",
   "metadata": {},
   "source": [
    "# get_read from bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e489a66b-70d1-46b1-a88d-cc0a4baa076b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f2608f6-a6c0-41a7-9fde-1b10420fbf22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('pep.fasta','w') as f:\n",
    "#     for sample in ['CRC-P59-T_2','CRC-P59-T_1','CRC-P67-T']:\n",
    "#         pepgenedict = dict(zip(snvresult[sample]['gene_ids'],snvresult[sample]['gene_name']))\n",
    "#         tmp = pep2snvresult[sample]\n",
    "#         tmp['gene'] = tmp['snv'].map(pepgenedict)\n",
    "#         tmp['sample'] = sample\n",
    "#         for idx,row in tmp.iterrows():\n",
    "#             pep,snv,gene,sample = row\n",
    "#             f.write(f'> {snv}|||{gene}|||{sample}\\n')\n",
    "#             f.write(f'{pep}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff4f7a7f-ddb4-4968-ba22-9d5ff6f6dd68",
   "metadata": {
    "tags": []
   },
   "source": [
    "## KRAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882524b4-91d7-4c7b-a866-a81403e9d497",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'CRC-P67-T'\n",
    "snv_name = \"chr12_25245350:C>A\"\n",
    "snv_proxy = snv_name.split(':')[0]\n",
    "chrom = snv_proxy.split('_')[0]\n",
    "pos = int(snv_proxy.split('_')[1])\n",
    "\n",
    "ref = snv_name.split(':')[1].split('>')[0]\n",
    "alt = snv_name.split(':')[1].split('>')[1]\n",
    "\n",
    "\n",
    "resultfrombam = []\n",
    "bam_file = f\"{root_path}/otherdata_for_figure/Figure6/subbam/{sample}/{snv_proxy}.bam\"\n",
    "inputpysam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "for i in inputpysam.pileup(chrom,pos-1,pos,truncate=True,min_mapping_quality=10): # 1-base\n",
    "    for pileupread in i.pileups:\n",
    "        if not pileupread.is_del and not pileupread.is_refskip and not pileupread.alignment.is_qcfail and not pileupread.alignment.is_duplicate:\n",
    "            read_base = pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            x = pileupread.alignment.get_tag('Cx')\n",
    "            y = pileupread.alignment.get_tag('Cy')\n",
    "            resultfrombam.append([x,y,read_base])\n",
    "resultfrombam = pd.DataFrame(resultfrombam,columns = ['x','y','base'])\n",
    "resultfrombam['x'] = resultfrombam['x']//100*100\n",
    "resultfrombam['y'] = resultfrombam['y']//100*100\n",
    "resultfrombam['DNB'] = 'DNB_' + resultfrombam['x'].map(str) + '_' + resultfrombam['y'].map(str)\n",
    "resultfrombam = resultfrombam.drop_duplicates(['DNB','base'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cffdc3ac-e1a0-438e-88a1-3ddb848a35b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "snv = snv_adata[sample]\n",
    "sub_snv = snv[snv.obs['cluster'] == 'Tumor'].copy()\n",
    "sub_snv.obs['x'] = sub_snv.obsm['spatial'][:,0]\n",
    "sub_snv.obs['y'] = sub_snv.obsm['spatial'][:,1]\n",
    "snv_df = sub_snv.obs[['x','y']].copy()\n",
    "points = snv_df[['x','y']].values\n",
    "hull = concave_hull.concave_hull(points,concavity = 0.01)\n",
    "s = 5\n",
    "fig,ax = plt.subplots(figsize = (8,8))\n",
    "sns.scatterplot(x = snv.obsm['spatial'][:,0],y = snv.obsm['spatial'][:,1],s = 10,edgecolor = None,color = '#EEEEEE',rasterized =True)\n",
    "sns.scatterplot(x = snv[snv.obs['cluster']=='Tumor'].obsm['spatial'][:,0],y = snv[snv.obs['cluster']=='Tumor'].obsm['spatial'][:,1],s = 10,alpha = 0.3,edgecolor = None,color = '#B4635C',rasterized =True)\n",
    "sns.scatterplot(x = hull[:,0],y = hull[:,1],s = s,edgecolor = None,color = 'black',markers='s',rasterized =True)\n",
    "sns.scatterplot(data = resultfrombam[resultfrombam['base'] == ref] ,x = 'x', y = 'y',color = '#479FDA',s = 100,edgecolor = 'black',rasterized =True)\n",
    "sns.scatterplot(data = resultfrombam[resultfrombam['base'] == alt] ,x = 'x', y = 'y',color = '#D75755',s = 100,edgecolor = 'black',rasterized =True)\n",
    "ax.invert_yaxis()\n",
    "# fig.savefig(f'./MainFigure/{sample}.snv.Tumor.pdf')\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34009446-e3a2-4e7b-9c5d-da4ebb68750d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'CRC-P59-T_2'\n",
    "snv_name = \"chr12_25245350:C>T\"\n",
    "snv_proxy = snv_name.split(':')[0]\n",
    "chrom = snv_proxy.split('_')[0]\n",
    "pos = int(snv_proxy.split('_')[1])\n",
    "\n",
    "ref = snv_name.split(':')[1].split('>')[0]\n",
    "alt = snv_name.split(':')[1].split('>')[1]\n",
    "\n",
    "\n",
    "\n",
    "resultfrombam = []\n",
    "bam_file = f\"{root_path}/otherdata_for_figure/Figure6/subbam/{sample}/{snv_proxy}.bam\"\n",
    "inputpysam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "for i in inputpysam.pileup(chrom,pos-1,pos,truncate=True,min_mapping_quality=10): # 1-base\n",
    "    for pileupread in i.pileups:\n",
    "        if not pileupread.is_del and not pileupread.is_refskip and not pileupread.alignment.is_qcfail and not pileupread.alignment.is_duplicate:\n",
    "            read_base = pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            x = pileupread.alignment.get_tag('Cx')\n",
    "            y = pileupread.alignment.get_tag('Cy')\n",
    "            resultfrombam.append([x,y,read_base])\n",
    "resultfrombam = pd.DataFrame(resultfrombam,columns = ['x','y','base'])\n",
    "resultfrombam['x'] = resultfrombam['x']//100*100\n",
    "resultfrombam['y'] = resultfrombam['y']//100*100\n",
    "resultfrombam['DNB'] = 'DNB_' + resultfrombam['x'].map(str) + '_' + resultfrombam['y'].map(str)\n",
    "resultfrombam = resultfrombam.drop_duplicates(['DNB','base'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e66c7b41-05bd-49e2-9919-43a516501c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "snv = snv_adata[sample]\n",
    "sub_snv = snv[snv.obs['cluster'] == 'Tumor'].copy()\n",
    "sub_snv.obs['x'] = sub_snv.obsm['spatial'][:,0]\n",
    "sub_snv.obs['y'] = sub_snv.obsm['spatial'][:,1]\n",
    "snv_df = sub_snv.obs[['x','y']].copy()\n",
    "import concave_hull\n",
    "points = snv_df[['x','y']].values\n",
    "hull = concave_hull.concave_hull(points,concavity = 0.01)\n",
    "\n",
    "fig,ax = plt.subplots(figsize = (8,8))\n",
    "sns.scatterplot(x = snv.obsm['spatial'][:,0],y = snv.obsm['spatial'][:,1],s = 10,edgecolor = None,color = '#EEEEEE',rasterized =True)\n",
    "sns.scatterplot(x = snv[snv.obs['cluster']=='Tumor'].obsm['spatial'][:,0],y = snv[snv.obs['cluster']=='Tumor'].obsm['spatial'][:,1],s = 10,alpha = 0.2,edgecolor = None,color = '#B4635C',rasterized =True)\n",
    "sns.scatterplot(x = hull[:,0],y = hull[:,1],s = 5,edgecolor = None,color = 'black',markers='s',rasterized =True)\n",
    "sns.scatterplot(data = resultfrombam[resultfrombam['base'] == ref] ,x = 'x', y = 'y',color = '#479FDA',s = 100,edgecolor = 'black',rasterized =True)\n",
    "sns.scatterplot(data = resultfrombam[resultfrombam['base'] == alt] ,x = 'x', y = 'y',color = '#D75755',s = 100,edgecolor = 'black',rasterized =True)\n",
    "ax.invert_yaxis()\n",
    "# fig.savefig(f'./MainFigure/{sample}.snv.Tumor.pdf')\n",
    "fig.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spatialsnv",
   "language": "python",
   "name": "spatialsnv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
